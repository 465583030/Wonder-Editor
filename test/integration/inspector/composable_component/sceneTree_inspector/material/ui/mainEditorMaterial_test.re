open Wonder_jest;

open Expect;

open Expect.Operators;

open Sinon;

open MainEditorMaterialType;

let _ =
  describe("MainEditorMaterial", () => {
    let sandbox = getSandboxDefaultVal();
    beforeEach(() => {
      sandbox := createSandbox();
      MainEditorSceneTool.initStateAndGlWithJob(
        ~sandbox,
        ~noWorkerJobRecord=
          NoWorkerJobConfigToolEngine.buildNoWorkerJobConfig(),
        (),
      );
      EventListenerTool.buildFakeDom()
      |> EventListenerTool.stubGetElementByIdReturnFakeDom;
    });
    afterEach(() => restoreSandbox(refJsObjToSandbox(sandbox^)));
    describe("test set currentSceneTreeNode", () => {
      beforeEach(() =>
        MainEditorSceneTool.createDefaultScene(
          sandbox,
          MainEditorSceneTool.setFirstBoxTobeCurrentSceneTreeNode,
        )
      );

      describe("test change material", () => {
        describe("test snapshot", () => {
          test("test show default light material component", () => {
            let component = BuildComponentTool.buildMaterial();

            component |> ReactTestTool.createSnapshotAndMatch;
          });
          test("test change to basic material component", () => {
            DirectorToolEngine.prepareAndInitAllEnginState();



            let component = BuildComponentTool.buildMaterial();
            let materialType = BasicMaterial |> convertMaterialTypeToInt;

            BaseEventTool.triggerComponentEvent(
              component,
              MainEditorMaterialTool.triggerChangeMaterialTypeEvent(
                materialType,
              ),
            );

            component |> ReactTestTool.createSnapshotAndMatch;
          });
        });

        describe("test logic", () => {
          beforeEach(() => DirectorToolEngine.prepareAndInitAllEnginState());

          test(
            "currentSceneTreeNode's default material should be lightMaterial",
            () => {
            let materialType =
              MainEditorMaterialUtils.getMaterialTypeByGameObject(
                GameObjectTool.unsafeGetCurrentSceneTreeNode(),
              )
              |> StateLogicService.getEngineStateToGetData;

            materialType |> expect == MainEditorMaterialType.LightMaterial;
          });

          describe(
            "test change currentSceneTreeNode's lightMaterial to basicMaterial",
            () => {
            test(
              "test currentSceneTreeNode's material component should be basicMaterial",
              () => {
              MainEditorMaterialTool.setMaterialTypeToBeBaiscMaterial();

              GameObjectComponentEngineService.hasBasicMaterialComponent(
                GameObjectTool.unsafeGetCurrentSceneTreeNode(),
              )
              |> StateLogicService.getEngineStateToGetData
              |> expect == true;
            });
            test(
              "test gameObject should move from lightMaterialRenderArray to basicMaterialRenderArray",
              () => {
                let (basicMaterialRenderCount, lightMaterialRenderCount) =
                  MeshRendererToolEngine.getAllRenderArrayCount();

                MainEditorMaterialTool.setMaterialTypeToBeBaiscMaterial();

                MeshRendererToolEngine.getAllRenderArrayCount()
                |>
                expect == (
                            basicMaterialRenderCount + 1,
                            lightMaterialRenderCount - 1,
                          );
              },
            );
          });
        });
      });
    });
  });