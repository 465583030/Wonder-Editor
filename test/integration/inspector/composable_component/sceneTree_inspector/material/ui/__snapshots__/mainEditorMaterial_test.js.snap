// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MainEditorMaterial test set currentSceneTreeNode test change material test snapshot test change to basic material component 1`] = `
<article
  className="wonder-material"
>
  <div
    className=""
  >
    <article
      className="wonder-select"
    >
      <span
        className=""
      >
        shader
      </span>
      <select
        onChange={[Function]}
        value="0"
      >
        <option
          value="0"
        >
          basic_material
        </option>
        <option
          value="1"
        >
          light_material
        </option>
      </select>
    </article>
  </div>
  <div
    className=""
  >
    <article
      className="wonder-basic-material"
    >
      <article
        className="wonder-material-color"
      >
        <div
          className=""
        >
          <span
            className=""
          >
            color : 
          </span>
          <span
            className=""
          >
            #ffffff
          </span>
          <button
            className=""
            onClick={[Function]}
          >
            pick color
          </button>
        </div>
      </article>
      <article
        className="wonder-material-texture"
      >
        <div
          className="texture_ground"
          onDragEnter={[Function]}
          onDragLeave={[Function]}
          onDragOver={[Function]}
          onDrop={[Function]}
          style={
            Object {
              "opacity": "1",
            }
          }
        />
        <span
          className=""
        >
          map : 
        </span>
        <img
          src="./public/img/null.jpg"
        />
        <button
          className="texture_remove"
          onClick={[Function]}
        >
          remove
        </button>
      </article>
    </article>
  </div>
</article>
`;

exports[`MainEditorMaterial test set currentSceneTreeNode test change material test snapshot test show default light material component 1`] = `
<article
  className="wonder-material"
>
  <div
    className=""
  >
    <article
      className="wonder-select"
    >
      <span
        className=""
      >
        shader
      </span>
      <select
        onChange={[Function]}
        value="1"
      >
        <option
          value="0"
        >
          basic_material
        </option>
        <option
          value="1"
        >
          light_material
        </option>
      </select>
    </article>
  </div>
  <div
    className=""
  >
    <article
      className="wonder-light-material"
    >
      <article
        className="wonder-material-color"
      >
        <div
          className=""
        >
          <span
            className=""
          >
            diffcuse color : 
          </span>
          <span
            className=""
          >
            #ffffff
          </span>
          <button
            className=""
            onClick={[Function]}
          >
            pick color
          </button>
        </div>
      </article>
      <article
        className="wonder-material-texture"
      >
        <div
          className="texture_ground"
          onDragEnter={[Function]}
          onDragLeave={[Function]}
          onDragOver={[Function]}
          onDrop={[Function]}
          style={
            Object {
              "opacity": "1",
            }
          }
        />
        <span
          className=""
        >
          diffuse map : 
        </span>
        <img
          src="./public/img/null.jpg"
        />
        <button
          className="texture_remove"
          onClick={[Function]}
        >
          remove
        </button>
      </article>
      <article
        className="wonder-light-shininess"
      >
        <article
          className="wonder-float-input"
        >
          <span
            className="component-label"
          >
            Shininess : 
          </span>
          <input
            className="input-component float-input"
            onBlur={[Function]}
            onChange={[Function]}
            type="text"
            value="32"
          />
        </article>
      </article>
    </article>
  </div>
</article>
`;
